Replace winit's synthetic events by our own key tracking

Azorlogh opened the pull request:
# Objective

Defocusing a window while a key is held (such as when Alt+tabbing), will currently send a key release on X11 and Windows. This is likely the behavior that most people expect.  
However it's synthetic events from winit are unimplemented for WASM and some other platforms. (See https://github.com/rust-windowing/winit/issues/1272).
While we can implement it upstream, there is also some doubt about the synthetic events API as a whole (https://github.com/rust-windowing/winit/issues/3543), so I propose to do it in bevy directly for now.

## Solution

This PR implements key tracking in bevy directly so we can synthesize our own key release events across all platforms.

Note regarding X11 specifically:
- On `main`, pressing a keyboard shortcut to unfocus the window (`Ctrl + Super + ArrowRight` in my case) will yield the following events:
```
Pressed Control
Pressed Super
Released Control
Released ArrowRight
Released Super
```
- On this branch, it will yield the following sequence:
```
Pressed Control
Pressed Super
Released Control
Released Super
```
To me the behavior of this branch is more expected than `main`, because `main` produces an `ArrowRight` release without producing a press first.

## Testing

Tested in WASM and X11 with
```rust
App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Update, |mut keys: EventReader<KeyboardInput>| {
            for ev in keys.read() {
                error!("received {ev:?}");
            }
        })
        .run();
```

RobWalt commented on crates/bevy_winit/src/system.rs:
```diff
@@ -428,3 +434,13 @@ pub(crate) fn changed_windows(
+#[derive(PartialEq, Eq, Hash)]
+pub struct PressedKey {
+    pub logical_key: winit::keyboard::Key,
+    pub physical_key: winit::keyboard::PhysicalKey,
```
Sorry, this might be a bit picky: What value does this field provide at the moment? As far as I can tell we would be fine without it since the `PhysicalKey` is already the key of the `HashMap`.

Azorlogh replied:
True, this is mostly a leftover :ok_hand: And turns out I don't need the location either

RobWalt reviewed (commented):
Code looks good to me!

NEON725 commented:
Will this fix the issue of the Alt key "sticking" when alt-tabbing out of the browser running a WASM build? Logic seems like it should so long as WASM builds correctly track window focus.

Azorlogh commented:
> Will this fix the issue of the Alt key "sticking" when alt-tabbing out of the browser running a WASM build? Logic seems like it should so long as WASM builds correctly track window focus.

Yep, this should fix it.

alice-i-cecile commented:
@richchurcher @SpecificProtagonist can I have your review over here please?

richchurcher commented:
@Azorlogh I think because #12372 got merged and touched some of the same areas, this is gonna have a few conflicts. Would you mind taking another look and updating the branch?

Azorlogh commented:
I updated it for #12372, though I have a couple concerns on reliability:
- As I wrote in https://github.com/bevyengine/bevy/pull/12372#issuecomment-2395067994, there are false positives when switching between bevy windows (the `check_keyboard_focus_lost` sometimes erroneously thinks we switched to a non-bevy window). After my changes this will mean extra release events will be emitted (though maybe that's not too a huge problem).
    - I worried it might reintroduce the bug in https://github.com/bevyengine/bevy/issues/12273, but at the same time I didn't manage to reproduce it so I'm not sure.
- Also, `check_keyboard_focus_lost` doesn't emit events in `Events<bevy_window::WindowEvent>`, so if the user listens to those events they will miss them. (I guess we can just emit them in both streams as a quick fix though)

alice-i-cecile commented:
> * As I wrote in [Fix window spawning triggering ButtonInput<KeyCode>::just_pressed/just_released #12372 (comment)](https://github.com/bevyengine/bevy/pull/12372#issuecomment-2395067994), there are false positives when switching between bevy windows (the `check_keyboard_focus_lost` sometimes erroneously thinks we switched to a non-bevy window). After my changes this will mean extra release events will be emitted (though maybe that's not too a huge problem).

Extra release events should generally be safe enough.

>   * I worried it might reintroduce the bug in [Opening a Window on `keyboard_input.just_pressed` makes it trigger again #12273](https://github.com/bevyengine/bevy/issues/12273), but at the same time I didn't manage to reproduce it so I'm not sure.

I think this is broadly fine.

> * Also, `check_keyboard_focus_lost` doesn't emit events in `Events<bevy_window::WindowEvent>`, so if the user listens to those events they will miss them. (I guess we can just emit them in both streams as a quick fix though)

Emitting it in both would be good.

Azorlogh commented:
> Extra release events should generally be safe enough

Fair. They are now emitted in both streams :+1:

Azorlogh commented:
@BenjaminBrienen I just updated the PR to main :+1: From my side I think it's ready

alice-i-cecile reviewed (approved):
Good comments and I much prefer having some cross-platform consistency here. Not thrilled to be doing this rather than winit but bugs are bugs.
